src/core/term.ml

```diff
@@ -309,66 +296,30 @@ end
 module SymSet = Set.Make(Sym)
 module SymMap = Map.Make(Sym)
 
-(** Basic management of meta variables. *)
-module Meta : sig
-  type t = meta
-  (** Type of metavariables. *)
-
-  val compare : t -> t -> int
-  (** Comparison function for metavariables. *)
-
-  val fresh : ?name:string -> term -> int -> t
-  (** [fresh ?name a n] creates a fresh metavariable of type [a] and arity [n]
-     with the optional name [name]. *)
-
-  val fresh_box : ?name:string -> tbox -> int -> t Bindlib.box
-  (** [fresh_box ?name a n] is the boxed counterpart of [fresh_meta]. It is
-      only useful in the rare cases where the type of a metavariable contains
-      a free term variable environement. This should only happens when scoping
-      the rewriting rules, use this function with care.  The metavariable is
-      created immediately with a dummy type, and the type becomes valid at
-      unboxing. The boxed metavariable should be unboxed at most once,
-      otherwise its type may be rendered invalid in some contexts. *)
-
-  val set : t -> tmbinder -> unit
-  (** [set m v] sets the value of the metavariable [m] to [v]. Note that no
-      specific check is performed, so this function may lead to cyclic
-      terms. *)
-
-  val name : t -> string
-  (** [name m] returns a string representation of [m]. *)
-
-  val reset_meta_counter : unit -> unit
-  (** [reset_counter ()] resets the counter used to produce meta keys. *)
-end = struct
+(** Sets and maps of metavariables. *)
+module Meta = struct
   type t = meta
-  let compare m1 m2 = m1.meta_key - m2.meta_key
-  let meta_counter : int Stdlib.ref = Stdlib.ref (-1)
-  let reset_meta_counter () = Stdlib.(meta_counter := -1)
-
-  let fresh : ?name:string -> term -> int -> t = fun ?name a n ->
-      { meta_key = Stdlib.(incr meta_counter; !meta_counter); meta_name = name
-      ; meta_type = ref a; meta_arity = n; meta_value = ref None }
-
-  let set : t -> tmbinder -> unit = fun m v ->
-    m.meta_type := Kind; (* to save memory *) m.meta_value := Some(v)
-
-  let name : t -> string = fun m ->
-    match m.meta_name with
-    | Some(n) -> n
-    | None    -> string_of_int m.meta_key
-
-  let fresh_box : ?name:string -> tbox -> int -> t Bindlib.box =
-    fun ?name a n ->
-    let m = fresh ?name Kind n in
-    Bindlib.box_apply (fun a -> m.meta_type := a; m) a
-
+  let compare m1 m2 = m2.meta_key - m1.meta_key
 end
 
-(** Sets and maps of metavariables. *)
 module MetaSet = Set.Make(Meta)
 module MetaMap = Map.Make(Meta)
 
+(** Representation of unification problems. *)
+type problem =
```

Comment:
What would you think of abstracting this datatype, since references may well be error-prone?

Reply:
What do you mean? Enforcing the use of update functions? Not sure it improves things very much. The `<-` are easy to track.

Reply:
Indeed it's what I had in mind. Alright.